lib/teleterm/daemon/daemon.go

```diff
@@ -215,55 +184,98 @@ func (s *Service) RemoveGateway(ctx context.Context, gatewayURI string) error {
 
 // removeGateway assumes that mu is already held by a public method.
 func (s *Service) removeGateway(gateway *gateway.Gateway) error {
+	// In case closing the gateway fails, we want to return early and leave the gateway in the map.
+	//
+	// This way the gateway will remain shown as available in the app. The user will see the error
+	// related to closing the gateway and will be able to attempt to close it again.
+	//
+	// This is preferable to removing the gateway from the hash map. Since Connect remembers the port
+	// used for a particular db server + username pair, the user would be able to reopen the same
+	// gateway on the occupied port only to see an error.
 	if err := gateway.Close(); err != nil {
 		return trace.Wrap(err)
 	}
 
-	// remove closed gateway from list
-	for index := range s.gateways {
-		if s.gateways[index] == gateway {
-			s.gateways = append(s.gateways[:index], s.gateways[index+1:]...)
-			return nil
-		}
-	}
+	delete(s.gateways, gateway.URI.String())
 
-	return trace.NotFound("gateway %v not found in gateway list", gateway.URI.String())
+	return nil
 }
 
 // RestartGateway stops a gateway and starts a new one with identical parameters.
 // It also keeps the original URI so that from the perspective of Connect it's still the same
 // gateway but with fresh certs.
 func (s *Service) RestartGateway(ctx context.Context, gatewayURI string) error {
-	gateway, err := s.FindGateway(gatewayURI)
+	oldGateway, err := s.FindGateway(gatewayURI)
 	if err != nil {
 		return trace.Wrap(err)
 	}
 
 	s.mu.Lock()
```

Comment:
Similar to my previous comment - taking out a separate lock after `FindGateway` makes it possible that:

- you successfully find the gateway to restart
- the gateway you found is removed/replaced
- you then call remove gateway on a gateway that was already removed